jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/TypeLayoutCheckHelper.java

```diff
@@ -0,0 +1,234 @@
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package jdk.internal.foreign.abi;
+
+import java.util.List;
+import java.lang.invoke.MethodType;
+
+import jdk.incubator.foreign.CLinker.TypeKind;
+import static jdk.incubator.foreign.CLinker.TypeKind.*;
+import jdk.incubator.foreign.FunctionDescriptor;
+import jdk.incubator.foreign.GroupLayout;
+import jdk.incubator.foreign.MemoryAddress;
+import jdk.incubator.foreign.MemoryLayout;
+import jdk.incubator.foreign.MemorySegment;
+import jdk.incubator.foreign.ValueLayout;
+
+/**
+ * The Class is a collection of helper methods to validate the method types
+ * against the corresponding layouts in the function descriptor.
+ */
+class TypeLayoutCheckHelper {
+
+	/* Verify whether the specified layout and the corresponding type are valid and match each other */
+	static void checkIfValidLayoutAndType(MethodType targetMethodType, MemoryLayout[] argumentLayouts, MemoryLayout returnLayout) {
+		Class<?> retType = targetMethodType.returnType();
+		if (!validateArgRetTypeClass(retType) && (retType != void.class)) {
+			throw new IllegalArgumentException("The return type must be primitive/void, MemoryAddress or MemorySegment" + ": retType = " + retType);  //$NON-NLS-1$ //$NON-NLS-2$
+		}
+
+		validateLayoutAgainstType(returnLayout, targetMethodType.returnType());
+
+		Class<?>[] argTypes = targetMethodType.parameterArray();
+		int argTypeCount = argTypes.length;
+		int argLayoutCount = argumentLayouts.length;
+		if (argTypeCount != argLayoutCount) {
+			throw new IllegalArgumentException("The arity (" + argTypeCount //$NON-NLS-1$
+					+ ") of the argument types is inconsistent with the arity ("  //$NON-NLS-1$
+					+ argLayoutCount + ") of the argument layouts");  //$NON-NLS-1$
+		}
+
+		for (int argIndex = 0; argIndex < argLayoutCount; argIndex++) {
+			if (!validateArgRetTypeClass(argTypes[argIndex])) {
+				throw new IllegalArgumentException("The passed-in argument type at index " + argIndex + " is neither primitive nor MemoryAddress nor MemorySegment"); //$NON-NLS-1$ //$NON-NLS-2$
+			}
+			validateLayoutAgainstType(argumentLayouts[argIndex], argTypes[argIndex]);
+		}
+	}
+
+	/* Verify whether the specified type is primitive, MemoryAddress (for pointer) or MemorySegment (for struct) */
+	private static boolean validateArgRetTypeClass(Class<?> targetType) {
+		if (!targetType.isPrimitive()
+		&& (targetType != MemoryAddress.class)
+		&& (targetType != MemorySegment.class)
+		) {
+			return false;
+		}
+		return true;
+	}
+
+	/* Check the validity of the layout against the corresponding type */
+	private static void validateLayoutAgainstType(MemoryLayout targetLayout, Class<?> targetType) {
+		boolean isPrimitiveLayout = false;
```

Comment:
this is unused

Reply:
Removed (leftover in the code at phase 1).